### PR TITLE
PLAT-65757: Move to a previous or next page via a channel up and down immediately

### PR DESCRIPTION
**Pull request by ybsung**
### Checklist

* [X] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [X] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [X] At least one test case is included for this feature or bug fix
* [ ] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)

The UX for page navigation via channel up and down was updated. VirtualList or Scroller jump to a previous or next page via a channel up and down immediately instead of jumping to the first or last visible item in the current page.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

There were 3 conditions to jump via channel up and down in VirtualList and Scroller with the previous UX.
Priority 1> If Spotlight was not on the first or last visible item in the current page, they jumped to it.
Priority 2> If Spotlight was on the first or last visible item in the current page and there was the spottable item in the previous or next page, it jumped to the first or last item in the page.
Priority 3> If Spotlight was on the first or last visible item in the current page and there was no the spottable item in the previous or next page, it jumped to the next spottable item.

For the update UX for page navigation, I just moved the Priority 1> to the lowest priority. So VirtualList and Scroll will jump to the previous or next page immediately via channel up and down.

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

Even though changing the priority, Scroller did not meet the full page navigation UX yet. So I also fixed it together.

### Comments

PLAT-65757

### CODE REVIEW

File: packages/moonstone/Scroller/Scroller.js [hunk 1]
```diff
@@ -298,49 +298,55 @@ class ScrollerBase extends Component {
 		}
 	}
 
-	getNextEndPoint = (direction, oSpotBounds) => {
-		const bounds = this.uiRefCurrent.getScrollBounds();
-
+	getNextEndPoint = (direction, scrollerBounds, focusedItemBounds, pageDiff) => {
```
Comment (0x64) [inline packages/moonstone/Scroller/Scroller.js]: I prefer a boolean `ignoreCurrentPage` or something than a number `pageDiff` for limitation for clearer understanding.

File: packages/moonstone/Scroller/Scroller.js [hunk 2]
```diff
@@ -298,49 +298,55 @@ class ScrollerBase extends Component {
...
-				oPoint.x = oSpotBounds.left - bounds.clientWidth;
-				oPoint.y = oSpotBounds.top;
+				oPoint.x = focusedItemBounds.left - scrollerBounds.width;
+				oPoint.y = focusedItemBounds.top;
 				break;
 			case 'down':
-				oPoint.x = oSpotBounds.left;
-				oPoint.y = oSpotBounds.top + oSpotBounds.height + bounds.clientHeight;
+				oPoint.x = focusedItemBounds.left;
+				oPoint.y = scrollerBounds.top + scrollerBounds.height * (pageDiff + 1);
```
Comment (0x64) [inline packages/moonstone/Scroller/Scroller.js]: What about to use scrollerBounds.bottom?
> oPoint.y = scrollerBounds.bottom + ignoreCurrentPage ? scrollerBounds.height : 0;

File: packages/moonstone/Scroller/Scroller.js [hunk 3]
```diff
@@ -298,49 +298,55 @@ class ScrollerBase extends Component {
...
 				break;
 			case 'right':
-				oPoint.x = oSpotBounds.left + oSpotBounds.width + bounds.clientWidth;
-				oPoint.y = oSpotBounds.top;
+				oPoint.x = focusedItemBounds.left + focusedItemBounds.width + scrollerBounds.width;
```
Comment (0x64) [inline packages/moonstone/Scroller/Scroller.js]: I'm really confused about the rule for this case here.

File: packages/moonstone/Scroller/Scroller.js [hunk 4]
```diff
@@ -298,49 +298,55 @@ class ScrollerBase extends Component {
...
+			scrollerBounds = this.uiRefCurrent.containerRef.current.getBoundingClientRect(),
+			focusedItemBounds = focusedItem.getBoundingClientRect();
+		let candidateNode = null, endPoint;
+
+		/* Find a spottable item in the current page */
+		endPoint = this.getNextEndPoint(direction, scrollerBounds, focusedItemBounds, 0);
+		const candidateNodeInCurrentPage = getTargetByDirectionFromPosition(reverseDirection, endPoint, spotlightId);
 
 		/* Find a spottable item in the next page */
+		endPoint = this.getNextEndPoint(direction, scrollerBounds, focusedItemBounds, 1);
```
Comment (0x64) [inline packages/moonstone/Scroller/Scroller.js]: If a candidate is found in the current page, do we need to run this line?

File: packages/moonstone/Scroller/Scroller.js [hunk 5]
```diff
@@ -298,49 +298,55 @@ class ScrollerBase extends Component {
...
 		candidateNode = getTargetByDirectionFromPosition(reverseDirection, endPoint, spotlightId);
 
 		/* Find a spottable item in a whole data */
-		if (candidateNode === focusedItem) {
+		if (candidateNode === focusedItem || candidateNode === candidateNodeInCurrentPage) {
```
Comment (0x64) [inline packages/moonstone/Scroller/Scroller.js]: When this if clause is true, what is the value of `endPoint`?

File: packages/moonstone/Scroller/Scroller.js [hunk 6]
```diff
@@ -298,49 +298,55 @@ class ScrollerBase extends Component {
...
 			candidateNode = getTargetByDirectionFromPosition(direction, endPoint, spotlightId);
 		}
 
-		/* If there is no spottable item next to the current item */
 		if (candidateNode === focusedItem) {
 			return null;
+		} else if (candidateNode === candidateNodeInCurrentPage) {
+			return candidateNodeInCurrentPage;
```
Comment (0x64) [inline packages/moonstone/Scroller/Scroller.js]: The returned value of this line is as same as `candidateNode`, so we will get the same result without this _else if_ clause.

File: packages/moonstone/Scrollable/Scrollable.js [hunk 7]
```diff
@@ -421,35 +421,15 @@ class ScrollableBase extends Component {
...
-			if (next !== focusedItem) {
-				Spotlight.focus(next);
-			/* 2. Find spottable item out of viewport */
-			// For Scroller
-			} else if (this.childRef.current.scrollToNextPage) {
-				next = this.childRef.current.scrollToNextPage({direction, focusedItem, reverseDirection: rDirection, spotlightId});
+			const spotlightId = containerRef.current.dataset.spotlightId;
+
+			if (this.childRef.current.scrollByPage) {
+				const next = this.childRef.current.scrollByPage({direction, focusedItem, reverseDirection: reverseDirections[direction], spotlightId});
```
Comment (0x64) [inline packages/moonstone/Scrollable/Scrollable.js]: `spotlightId` is used just once. We don't need to define the const variable.

**Comment by 0x64** [comment]
I have two general questions. The first one: Is removing a searching logic to find the target in the current page from `scrollByPage` in _Scrollable_ intended? The second one: do this PR affect any behaviors on overscroll effects? (overscroll logic depends on the result of `scrollByPage` method.)

File: packages/moonstone/Scroller/Scroller.js (hunk 1 above)
Comment (ybsung) [reply packages/moonstone/Scroller/Scroller.js]: I replaced `pageDiff` with `isInNextPage`.

File: packages/moonstone/Scroller/Scroller.js (hunk 2 above)
Comment (ybsung) [reply packages/moonstone/Scroller/Scroller.js]: I agree with your opinion.

File: packages/moonstone/Scroller/Scroller.js (hunk 3 above)
Comment (ybsung) [reply packages/moonstone/Scroller/Scroller.js]: We don't need the `right` and `left` cases. So I removed those cases.

File: packages/moonstone/Scroller/Scroller.js (hunk 4 above)
Comment (ybsung) [reply packages/moonstone/Scroller/Scroller.js]: Yes. We need to find the spottable item in the next page or after the next page first.

File: packages/moonstone/Scroller/Scroller.js (hunk 5 above)
Comment (ybsung) [reply packages/moonstone/Scroller/Scroller.js]: The code was not clear. So I cleaned up.

File: packages/moonstone/Scroller/Scroller.js (hunk 6 above)
Comment (ybsung) [reply packages/moonstone/Scroller/Scroller.js]: That's true. I cleaned up the return cases again.

File: packages/moonstone/Scrollable/Scrollable.js (hunk 7 above)
Comment (ybsung) [reply packages/moonstone/Scrollable/Scrollable.js]: I'll fix it.

**Comment by ybsung** [comment]
@0x64 
>The first one: Is removing a searching logic to find the target in the current page from scrollByPage in Scrollable intended?

We need to search the spottable item in the next page first. So I remove the logic that you mentioned.

> The second one: do this PR affect any behaviors on overscroll effects? (overscroll logic depends on the result of scrollByPage method.)

I didn't change the return value for `scrollByPage` in `Scrollable`. So the behaviors won't be changed.

**Comment by ryanjduffy** [comment]
Closing in favor of #2262

**Pull request closed without merging**